### PR TITLE
FEA-3206: Fix NormalFormalParameters

### LINKED ISSUE #110: Incorrect indexing referencing `this.x` in constructor
This dart code
```dart
class Foo {
    final String bar;
    Foo(this.bar) {}
}
```

Results in this indexing snapshot:
```dart
  class Foo {
    final String bar;
//               ^^^ definition scip-dart pub dart_test 1.0.0 lib/main.dart/Foo#bar.
  
    Foo(this.bar) {}
//      ^^^^ reference scip-dart pub dart_test 1.0.0 lib/main.dart/Foo#bar.
//           ^^^ definition local 0
  }
```

instead of treating `this` as the bar symbol, `bar` should refer to bar

For parity with the dart language server, `this` should not be indexed, eg:

```dart
    Foo(this.bar) {}
//           ^^^ reference scip-dart pub dart_test 1.0.0 lib/main.dart/Foo#bar.
```

### PR DESCRIPTION
**Pull request by matthewnitschke-wk**
# [FEA-3206](https://jira.atl.workiva.net/browse/FEA-3206)
Issue Status

Closes #110 

Previously constructor parameters would incorrectly use `this` as the symbol for the reference. This pr addresses that problem

There was another problem where parameters would also, _always_ declare a "definition". Parameters that are referencing an instance variable (`Foo(this.bar)`), are not a definition, just a reference. This pr also fixes that problem by not marking these as definitions

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: snapshots/output/basic-project/lib/more.dart [hunk 1]
```diff
@@ -35,6 +35,7 @@
   class Animal with SleepMixin {
 //      ^^^^^^ definition scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#
 //      documentation ```dart
+//      relationship scip-dart pub dart_test 1.0.0 lib/more.dart/SleepMixin# implementation
```
Comment (matthewnitschke-wk) [inline snapshots/output/basic-project/lib/more.dart:38]: we're running relationship indexing on `basic-project` now, snapshots were not previously generated due to failing CI that neglected to run on previous PRs

File: snapshots/output/basic-project/lib/more.dart [hunk 2]
```diff
@@ -53,12 +54,8 @@
...
 //  documentation ```dart
 //  ^^^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#
-//         ^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#name.
-//              ^^^^ definition local 0
-//              documentation ```dart
-//                              ^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#type.
-//                                   ^^^^ definition scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#<constructor>().(type)
-//                                   documentation ```dart
+//              ^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#name.
+//                                   ^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#type.
```
Comment (matthewnitschke-wk) [inline snapshots/output/basic-project/lib/more.dart:58]: Few changes here:

- This is the actual fix that the PR was doing, no longer referring to `this` as the declaration
    ```dart
    Animal(this.name, {required this.type}) {
    //     ^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#name.
    ```
    becomes
    ```dart
    Animal(this.name, {required this.type}) {
    //          ^^^^ reference scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#name.
    ```

- `^^^^ definition local 0` on `name` was removed - `name` is not actually a definition, its simply a reference to the name var on the class
- `definition scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#<constructor>().(type)` was removed. Similar to `name`, there is not actual "definition" here, its just a reference

File: snapshots/output/basic-project/lib/more.dart [hunk 3]
```diff
@@ -102,6 +99,7 @@
 //  ^^^^^^ reference scip-dart pub dart:core 2.19.0 dart:core/string.dart/String#
 //         ^^^^^^^^ definition scip-dart pub dart_test 1.0.0 lib/more.dart/Animal#toString().
 //         documentation ```dart
+//         relationship scip-dart pub dart:core 2.19.0 dart:core/object.dart/Object#toString(). implementation reference
```
Comment (matthewnitschke-wk) [inline snapshots/output/basic-project/lib/more.dart:102]:
```dart
String toString
```

is overriding here, relationship is accurately noting this

File: snapshots/output/basic-project/lib/more.dart [hunk 4]
```diff
@@ -114,17 +112,17 @@
 //    documentation ```dart
 //                 ^^^^ reference scip-dart pub dart:core 2.19.0 dart:core/list.dart/List#
 //                      ^^^ reference scip-dart pub dart:core 2.19.0 dart:core/int.dart/int#
-//                           ^^^^^^^ definition local 1
+//                           ^^^^^^^ definition local 0
```
Comment (matthewnitschke-wk) [inline snapshots/output/basic-project/lib/more.dart:115]: the local variable changes here are just due to no longer declaring `local 0` above, no cause for concern

File: snapshots/output/basic-project/lib/other.dart [hunk 5]
```diff
@@ -26,34 +26,30 @@
 //  ^^^ definition scip-dart pub dart_test 1.0.0 lib/other.dart/Foo#<constructor>().
 //  documentation ```dart
 //  ^^^ reference scip-dart pub dart_test 1.0.0 lib/other.dart/Foo#
-//      ^^^^ reference local 0
-//           ^^^^ definition local 1
-//           documentation ```dart
+//           ^^^^ reference local 0
```
Comment (matthewnitschke-wk) [inline snapshots/output/basic-project/lib/other.dart:29]: We were previously always calculating a definition for parameters, even formal ones referring to instance variables, now we aren't, hence the removal of `definition local 1`

`_far` is a private instance variable, so the reference is just a `local 0`

File: snapshots/output/basic-project/lib/relationships.dart [hunk 6]
```diff
@@ -1,54 +1,54 @@
   abstract class Mammal {
-// definition scip-dart pub dart_test 1.0.0 lib/main.dart/
-//               ^^^^^^ definition scip-dart pub dart_test 1.0.0 lib/main.dart/Mammal#
+// definition scip-dart pub dart_test 1.0.0 lib/relationships.dart/
+//               ^^^^^^ definition scip-dart pub dart_test 1.0.0 lib/relationships.dart/Mammal#
```
Comment (matthewnitschke-wk) [inline snapshots/output/basic-project/lib/relationships.dart:3]: the file was moved/renamed, these changes are just updating for that difference

**Review by todbachman-wf — APPROVED** [review]

**Review by todbachman-wf — APPROVED** [review]

**Comment by matthewnitschke-wk** [comment]
QA +1

- [x]  CI does indeed pass

🚀 @Workiva/release-management-p 🚢

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**